### PR TITLE
Enable removing of initializers 

### PR DESCRIPTION
**Pull request by martinrybak**
Thanks for this great library! I am trying to use it for unit tests, but I need the lifetime for initializers to be per test class, not the lifetime of the test run. To do that, I would like to be able to remove initializers from the registry by calling:

```[[BMInitializerRegistry injectableRegistry] removeAllInitializers];```

Do you recommend a different solution?

### CODE REVIEW

**Comment by 0xc010d** [comment]
Nice pull request!

Though there's a question. Do you really need to remove all initializers at once? 

As the first step I'd propose to add ```- (void)unregisterInitializer;``` method to ```BMInitializer```.
If we consider moving registries to public, then it's needed to do that for all of them, not only for ```+ (instancetype)injectableRegistry;```.

**Comment by martinrybak** [comment]
Thanks for the reply! The idea is that I'd like to start with a clean slate for every test, so removing all initializers at once would be ideal. Calling an ```unregisterInitializer``` method on each ```BMInitializer``` instance would work but requires keeping a pointer alive to each.

Maybe there is a better way to accomplish a global "reset"?

**Comment by AlexDenisov** [comment]
Hi @martinrybak, sorry for the long delay.
Overall looks good, thought I'd suggest to keep `removeInitializer` and `removeAllInitializers` as you proposed, but also add `unregisterInitializer` as @0xc010d suggested (just for consistency of API).
Also, I'd keep `BMInitializerRegistry+Injectable.h` private, because you can obtain a reference to it from `BMInitializer` itself - this way we don't need to make all the registries public (though we have only one such registry).

Also, there is `BMInjectableInitializerFinder`, which needs to be invalidated after removing of initializer. It's possible to just find and remove cached initializer, but seems it will be easier and less error prone to just cleanup all of them, since the purpose of these methods (I suppose) only for testing.

And last but not least - I'd really appreciate if you write tests for all the changes you make ;)

Please, let me know if you need any help and/or support implementing this feature.

**Comment by martinrybak** [comment]
OK thanks for the guidance!

**Comment by martinrybak** [comment]
Sorry, I haven't used Cedar before. How do I get the tests to run?
screen shot 2015-09-22 at 1 24 32 pm

**Comment by AlexDenisov** [comment]
@martinrybak you just need to select 'Specs' target and run it (Cmd + R) using some simulator.

**Comment by martinrybak** [comment]
Thanks. BTW here is a simple example of the way that I'm using BloodMagic in my ```XCTestCases```:

```
- (void)setUp
{
	[super setUp];
	[BMInitializer addInjectableInitializerForClass:[NSNotificationCenter class] block:^id{
		return [NSNotificationCenter defaultCenter];
	}];
}

- (void)tearDown
{
	[BMInitializer removeAllInitializers];
	[super tearDown];
}
```

This example happens to use a few of my custom class categories on ```BMInitializer``` for brevity:

```
+ (void)addInjectableInitializerForClass:(Class)propertyClass block:(id(^)(void))block
{
	BMInitializer* initializer = [BMInitializer injectableInitializer];
	initializer.propertyClass = propertyClass;
	if (containerClass) {
		initializer.containerClass = containerClass;
	}
	initializer.initializer = ^id (id sender){
		return block();
	};
	[initializer registerInitializer];
}

+ (void)removeAllInitializers
{
	[[BMInitializerRegistry injectableRegistry] removeAllInitializers];
}
```

I would love for there to be a way to access the ```[BMInitializerRegistry injectableRegistry] ``` to call ```removeAllInitializers``` without going through an instance of ```BMInitializer```.

**Comment by AlexDenisov** [comment]
```[[BMInitializerRegistry injectableRegistry] removeAllInitializers];```

I don't really mind, let's go with it :+1: 

The only thing left: tests ;)

Also, it would be nice if you will squash[1] commits after you finish.

[1] https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history

**Comment by martinrybak** [comment]
Will write tests and squash, thanks!

**Comment by martinrybak** [comment]
@AlexDenisov I took a pass at writing relevant tests. I look forward to your comments. Thanks!

File: Specs/Specs/Modules/Injectable/BMInitializerFinderSpec.mm [hunk 1]
```diff
@@ -36,7 +42,28 @@
...
+			
+			BMInjectableInitializerFinder *finder = [BMInjectableInitializerFinder finder];
+			magic_initializer_t initializer_t = [finder initializerForProperty:property];
+
+			initializer_t == nil should_not be_truthy;
+			
+			[finder invalidate];
+			initializer_t = [finder cachedInitializerForProperty:property];
+			
+			initializer_t == nil should be_truthy;
```
Comment (AlexDenisov) [inline Specs/Specs/Modules/Injectable/BMInitializerFinderSpec.mm]: Cedar has `nil` matcher:
```
initializer should_not be_nil;
// ...
initializer should be_nil;
```

File: Specs/Specs/Modules/Injectable/BMInitializerRegistrySpecs.mm [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			finder should have_received(NSSelectorFromString(@"invalidate"));
+		});
+		it(@"should invalidate the cache when removing all initializers", ^{
+			BMInjectableInitializerFinder* finder = [BMInjectableInitializerFinder finder];
+			spy_on(finder);
+			
+			[registry addInitializer:[BMInitializer new]];
+			[registry removeAllInitializers];
+			
+			finder should have_received(NSSelectorFromString(@"invalidate"));
```
Comment (AlexDenisov) [inline Specs/Specs/Modules/Injectable/BMInitializerRegistrySpecs.mm]: Great :+1: !

File: Specs/Specs/Modules/Injectable/BMInitializerFinderSpec.mm (hunk 1 above)
Comment (martinrybak) [inline Specs/Specs/Modules/Injectable/BMInitializerFinderSpec.mm]: I had tried that but got this error:
```Attempt to compare non-pointer type to nil```

Comment (AlexDenisov) [inline Specs/Specs/Modules/Injectable/BMInitializerFinderSpec.mm]: I see, yes, it's because of old version of Cedar...

**Pull request merged**